### ISSUE TITLE
Replication of course updates may lead to inconsistent TrackedRace structures

### ISSUE THREAD

**Issue by aksajhfduwafe** [aksajhfduwafe/bugzilla-to-github-issues-test#3006]
| Field | Value |
|---|---|
| **Reporter** | @axeluhl |
| **Assignee** | @axeluhl |
| **Product** | Sailing Race Analytics |
| **Component** | Server |
| **Version** | unspecified |
| **Hardware** | All |
| **OS** | All |
| **Severity** | normal |
| **Priority** | P1 |
| **CC (unmapped)** | Frederik Petersen (`bugz@fpetersen.eu`), Simon Pamies (`s.pamies@the-placeholder-company.com`) |
| **Created** | 2015-06-24T11:25:51Z |
| **Original Bug** | [Bug 3006](https://bugzilla.sapsailing.com/bugzilla/show_bug.cgi?id=3006) |

**Duplicate of:** #3249

---

We've seen a strange sequence of exceptions on a replica during Kieler Woche 2015. This happened shortly after a course update was injected into the master where it was processed properly. Also, only one of the two replicas ran into the issue. A first exception looked like this:

INFO: Adding waypoint I-Rep to course 'R8-M Medal (470 M)'
Jun 24, 2015 10:04:48 AM com.sap.sailing.domain.base.impl.CourseImpl addWaypoint
INFO: Waypoint I-Rep added to course 'R8-M Medal (470 M)', before notifying listeners
Jun 24, 2015 10:04:48 AM com.sap.sailing.domain.base.impl.CourseImpl addWaypoint
INFO: Waypoint I-Rep added to course 'R8-M Medal (470 M)', after notifying listeners
Jun 24, 2015 10:04:48 AM com.sap.sse.util.SmartFutureCache$FutureTaskWithCancelBlocking call
SEVERE: SmartFutureCache.FutureTaskWithCancelBlocking.call
java.lang.NullPointerException
        at com.sap.sailing.domain.tracking.impl.TrackedLegImpl.getReferenceTimePoint(TrackedLegImpl.java:428)
        at com.sap.sailing.domain.tracking.impl.TrackedLegImpl.getWindwardDistanceFromLegStart(TrackedLegImpl.java:363)
        at com.sap.sailing.domain.ranking.AbstractRankingMetric.getWindwardDistanceTraveled(AbstractRankingMetric.java:617)
        at com.sap.sailing.domain.ranking.AbstractRankingMetric.getWindwardDistanceTraveled(AbstractRankingMetric.java:587)
        at com.sap.sailing.domain.ranking.OneDesignRankingMetric$1$1.getWindwardDistanceSailed(OneDesignRankingMetric.java:102)
        at com.sap.sailing.server.impl.RacingEventServiceImpl.lambda$1(RacingEventServiceImpl.java:3223)
        at com.sap.sailing.server.impl.RacingEventServiceImpl$$Lambda$45/2072163922.compare(Unknown Source)
        at java.util.TimSort.binarySort(TimSort.java:292)
        at java.util.TimSort.sort(TimSort.java:217)
        at java.util.Arrays.sort(Arrays.java:1512)
        at java.util.ArrayList.sort(ArrayList.java:1454)
        at com.sap.sailing.server.impl.RacingEventServiceImpl.getCompetitorInOrderOfWindwardDistanceTraveledFarthestFirst(RacingEventServiceImpl.java:3222)
        at com.sap.sailing.gwt.ui.server.SailingServiceImpl.computeQuickRanks(SailingServiceImpl.java:2063)
        at com.sap.sailing.gwt.ui.server.QuickRanksLiveCache$1.computeCacheUpdate(QuickRanksLiveCache.java:85)
        at com.sap.sailing.gwt.ui.server.QuickRanksLiveCache$1.computeCacheUpdate(QuickRanksLiveCache.java:1)
        at com.sap.sse.util.SmartFutureCache$FutureTaskWithCancelBlocking.call(SmartFutureCache.java:320)
        at com.sap.sse.util.SmartFutureCache$SettableCallable.call(SmartFutureCache.java:200)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
        
Later, the following exceptions were observed:

Jun 24, 2015 10:09:24 AM com.sap.sailing.domain.leaderboard.impl.ScoreCorrectionImpl getCorrectedNonMaxedScore
SEVERE: Exception while computing corrected non maxed score
java.lang.NullPointerException
        at com.sap.sailing.domain.tracking.impl.TrackedRaceImpl.getCurrentLeg(TrackedRaceImpl.java:1342)
        at com.sap.sailing.domain.tracking.impl.AbstractRaceRankComparator.compare(AbstractRaceRankComparator.java:75)
        at com.sap.sailing.domain.tracking.impl.AbstractRaceRankComparator.compare(AbstractRaceRankComparator.java:1)
        at java.util.TimSort.binarySort(TimSort.java:292)
        at java.util.TimSort.sort(TimSort.java:217)
        at java.util.Arrays.sort(Arrays.java:1512)
        at java.util.ArrayList.sort(ArrayList.java:1454)
        at java.util.Collections.sort(Collections.java:175)
        at com.sap.sailing.domain.tracking.impl.TrackedRaceImpl.getCompetitorsFromBestToWorst(TrackedRaceImpl.java:1235)
        at com.sap.sailing.domain.tracking.impl.TrackedRaceImpl.getCompetitorsFromBestToWorst(TrackedRaceImpl.java:1202)
        at com.sap.sailing.domain.tracking.impl.TrackedRaceImpl.getRank(TrackedRaceImpl.java:1195)
        at com.sap.sailing.domain.leaderboard.impl.AbstractLeaderboardImpl.getTrackedRank(AbstractLeaderboardImpl.java:109)
        at com.sap.sailing.domain.leaderboard.impl.AbstractSimpleLeaderboardImpl$1.call(AbstractSimpleLeaderboardImpl.java:436)
        at com.sap.sailing.domain.leaderboard.impl.AbstractSimpleLeaderboardImpl$1.call(AbstractSimpleLeaderboardImpl.java:1)
        at com.sap.sailing.domain.leaderboard.impl.ScoreCorrectionImpl.getCorrectedNonMaxedScore(ScoreCorrectionImpl.java:451)
        at com.sap.sailing.domain.leaderboard.impl.ScoreCorrectionImpl.getCorrectedScore(ScoreCorrectionImpl.java:366)
        at com.sap.sailing.domain.leaderboard.impl.AbstractSimpleLeaderboardImpl.getNetPoints(AbstractSimpleLeaderboardImpl.java:433)
        at com.sap.sailing.domain.leaderboard.impl.AbstractSimpleLeaderboardImpl.getCompetitorsFromBestToWorst(AbstractSimpleLeaderboardImpl.java:556)
        at com.sap.sailing.domain.leaderboard.impl.AbstractSimpleLeaderboardImpl.lambda$0(AbstractSimpleLeaderboardImpl.java:1209)
        at com.sap.sailing.domain.leaderboard.impl.AbstractSimpleLeaderboardImpl$$Lambda$39/1260994182.call(Unknown Source)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
        
The code at the point where the NPE occurs reads like thsi:

                        result = getTrackedLegStartingAt(waypointPassedLastAtOrBeforeTimePoint).getTrackedLeg(competitor);
                        
This suggests that there is no tracked leg found in the TrackedRace that starts at the waypoint given.

Could the replication of an UpdateMarkPassings operation be the reason? It contains a sequence of mark passings for a competitor. The replication operation is generated after the course lock has already been released on the master. There is no guarantee that when the operation is created the course still contains all waypoints to which the mark passings point. However, the replica would log an updateMarkPassings with a mark passing referring to a non-existing waypoint as an error, and we don't see such a log entry.

Needs closer investigation.

**Comment by aksajhfduwafe** [comment]
**@axeluhl** commented:

AbstractRankingMetric:608 has:

            for (final TrackedLeg trackedLeg : getTrackedRace().getTrackedLegs()) {

I cannot see that there is any course read lock obtained at this point. TrackedRaceImpl.getTrackedLegs() returns a copy of the tracked legs list. This would explain how a tracked leg could be analyzed further for which no waypoint exists. The easy fix for this exception therefore is to obtain the read lock for the course before starting this iteration.

It does most likely not explain the other exceptions that occurred and persisted.
